reader.go

```diff
@@ -29,31 +66,847 @@ type Reader struct {
+		return nil
+	}
+
+	generationID, memberID := r.membership()
+	request := offsetCommitRequestV3{
+		GroupID:       r.config.GroupID,
+		GenerationID:  generationID,
+		MemberID:      memberID,
+		RetentionTime: int64(r.config.RetentionTime / time.Millisecond),
+	}
+
+	for topic, partitions := range offsetsByTopicAndPartition {
+		t := offsetCommitRequestV3Topic{Topic: topic}
+		for partition, offset := range partitions {
+			t.Partitions = append(t.Partitions, offsetCommitRequestV3Partition{
+				Partition: int32(partition),
+				Offset:    offset,
+			})
+		}
+		request.Topics = append(request.Topics, t)
+	}
+
+	if _, err := conn.offsetCommit(request); err != nil {
+		r.withErrorLogger(func(l *log.Logger) {
+			l.Printf("unable to commit offsets for group, %v: %v", r.config.GroupID, err)
+		})
+	}
+
+	r.withLogger(func(l *log.Logger) {
+		l.Printf("committed offsets: %v", offsetsByTopicAndPartition)
+	})
+	return nil
+}
+
+func (r *Reader) commitLoopImmediate(conn offsetCommitter, stop <-chan struct{}) {
+	for {
+		select {
+		case <-stop:
+			return
+
+		case m, ok := <-r.commits:
+			if !ok {
+				r.withErrorLogger(func(l *log.Logger) {
+					l.Println("reader commit channel unexpectedly closed")
+				})
+				return
+			}
+
+			offsetsByTopicAndPartition := map[string]map[int]int64{
+				m.Topic: {m.Partition: m.Offset},
+			}
+
+			if err := r.commitOffsets(conn, offsetsByTopicAndPartition); err != nil {
+				r.withErrorLogger(func(l *log.Logger) {
+					l.Printf("unable to commit offset: %v", err)
+				})
+				return
+			}
+		}
+	}
+}
+
+func (r *Reader) commitLoopInterval(conn offsetCommitter, stop <-chan struct{}) {
+	ticker := time.NewTicker(r.config.HeartbeatInterval)
+	defer ticker.Stop()
+
+	offsetsByTopicAndPartition := map[string]map[int]int64{}
+
+	defer func() {
+		// commits any outstanding offsets on close
+		r.commitOffsets(conn, offsetsByTopicAndPartition)
+	}()
+
+	for {
+		select {
+		case <-stop:
+			return
+
+		case <-ticker.C:
+			if len(offsetsByTopicAndPartition) == 0 {
+				continue
+			}
+
+			if err := r.commitOffsets(conn, offsetsByTopicAndPartition); err != nil {
+				r.withErrorLogger(func(l *log.Logger) {
+					l.Printf("unable to commit offset: %v", err)
+				})
+				return
+			}
+			offsetsByTopicAndPartition = map[string]map[int]int64{}
+
+		case m, ok := <-r.commits:
+			if !ok {
+				r.withErrorLogger(func(l *log.Logger) {
+					l.Println("reader commit channel unexpectedly closed")
+				})
+				return
+			}
+
+			offsetsByPartition, ok := offsetsByTopicAndPartition[m.Topic]
+			if !ok {
+				offsetsByPartition = map[int]int64{}
+				offsetsByTopicAndPartition[m.Topic] = offsetsByPartition
+			}
+
+			if offset, ok := offsetsByPartition[m.Partition]; ok && offset >= m.Offset {
+				// offsets should only increase
+				continue
+			}
+
+			offsetsByPartition[m.Partition] = m.Offset
+		}
+	}
+}
+
+func (r *Reader) commitLoop(stop <-chan struct{}, done chan<- struct{}) {
```

Comment:
Instead of a `stop` channel here we could use a context, that way it could inherit from `r.stctx` and we would get cancellation propagation if either the base context is canceled or only the one we created for the commit loop.

The same logic applies for the heartbeat loop.